core/domain/state_domain_test.py

```diff
@@ -637,6 +639,587 @@ def test_validate_state_solicit_answer_details(self):
         self.assertEqual(init_state.solicit_answer_details, True)
 
+    def test_validate_solution_answer_is_exclusive(self):
+        exploration = self.save_new_valid_exploration('exp_id', 'owner_id')
+
+        # Solution should be set to None as default.
+        self.assertEqual(exploration.init_state.interaction.solution, None)
+
+        hints_list = []
+        solution = {
+            'answer_is_exclusive': False,
+            'correct_answer': 'hello_world!',
+            'explanation': {
+                'content_id': 'solution',
+                'html': 'hello_world is a string'
+            }
+        }
+        hints_list.append({
+            'hint_content': {
+                'content_id': 'hint_1',
+                'html': ''
+            },
+        })
+        exploration.init_state.update_interaction_hints(hints_list)
+        exploration.init_state.update_interaction_solution(solution)
+        exploration.validate()
+
+        solution = {
+            'answer_is_exclusive': 1,
+            'correct_answer': 'hello_world!',
+            'explanation': {
+                'content_id': 'solution',
+                'html': 'hello_world is a string'
+            }
+        }
+
+        exploration.init_state.update_interaction_solution(solution)
+        with self.assertRaisesRegexp(
+            Exception, 'Expected answer_is_exclusive to be bool, received 1'):
+            exploration.validate()
+
+    def test_validate_non_list_param_changes(self):
+        exploration = self.save_new_valid_exploration('exp_id', 'owner_id')
+        exploration.init_state.param_changes = 0
+
+        with self.assertRaisesRegexp(
+            Exception, 'Expected state param_changes to be a list, received 0'):
+            exploration.init_state.validate(None, True)
+
+    def test_validate_duplicate_content_id_with_answer_groups(self):
+        exploration = self.save_new_valid_exploration('exp_id', 'owner_id')
+        answer_groups_dict = {
+            'outcome': {
+                'dest': exploration.init_state_name,
+                'feedback': {
+                    'content_id': 'feedback_1',
+                    'html': 'Feedback'
+                },
+                'labelled_as_correct': False,
+                'param_changes': [],
+                'refresher_exploration_id': None,
+                'missing_prerequisite_skill_id': None
+            },
+            'rule_specs': [{
+                'inputs': {
+                    'x': 'Test'
+                },
+                'rule_type': 'Contains'
+            }],
+            'training_data': [],
+            'tagged_misconception_id': None
+        }
+
+        exploration.init_state.update_interaction_answer_groups(
+            [answer_groups_dict])
+        exploration.init_state.update_content({
+            'content_id': 'feedback_1',
+            'html': 'Feedback'
+        })
+
+        with self.assertRaisesRegexp(Exception, 'Found a duplicate content id'):
```

Comment:
Add full log with the content_id which is duplicate (this is the main part of this test, right?) 
Ditto everywhere below.

Reply:
done